transports/webrtc-websys/src/stream/poll_data_channel.rs

```diff
@@ -0,0 +1,225 @@
+
+    /// Waker for when we are waiting to write (again) to the DC because we previously exceeded the [`MAX_MSG_LEN`] threshold.
+    write_waker: Rc<AtomicWaker>,
+
+    /// Waker for when we are waiting for the DC to be closed.
+    close_waker: Rc<AtomicWaker>,
+
+    // Store the closures for proper garbage collection.
+    // These are wrapped in an [`Rc`] so we can implement [`Clone`].
+    _on_open_closure: Rc<Closure<dyn FnMut(RtcDataChannelEvent)>>,
+    _on_write_closure: Rc<Closure<dyn FnMut(Event)>>,
+    _on_close_closure: Rc<Closure<dyn FnMut(Event)>>,
+    _on_message_closure: Rc<Closure<dyn FnMut(MessageEvent)>>,
+}
+
+impl PollDataChannel {
+    pub(crate) fn new(inner: RtcDataChannel) -> Self {
+        let open_waker = Rc::new(AtomicWaker::new());
+        let on_open_closure = Closure::new({
+            let open_waker = open_waker.clone();
+
+            move |_: RtcDataChannelEvent| {
+                log::trace!("DataChannel opened");
+                open_waker.wake();
+            }
+        });
+        inner.set_onopen(Some(on_open_closure.as_ref().unchecked_ref()));
+
+        let write_waker = Rc::new(AtomicWaker::new());
+        inner.set_buffered_amount_low_threshold(0);
+        let on_write_closure = Closure::new({
+            let write_waker = write_waker.clone();
+
+            move |_: Event| {
+                log::trace!("DataChannel available for writing (again)");
+                write_waker.wake();
+            }
+        });
+        inner.set_onbufferedamountlow(Some(on_write_closure.as_ref().unchecked_ref()));
+
+        let close_waker = Rc::new(AtomicWaker::new());
+        let on_close_closure = Closure::new({
+            let close_waker = close_waker.clone();
+
+            move |_: Event| {
+                log::trace!("DataChannel closed");
+                close_waker.wake();
+            }
+        });
+        inner.set_onclose(Some(on_close_closure.as_ref().unchecked_ref()));
+
+        let new_data_waker = Rc::new(AtomicWaker::new());
+        let read_buffer = Rc::new(Mutex::new(BytesMut::new())); // We purposely don't use `with_capacity` so we don't eagerly allocate `MAX_READ_BUFFER` per stream.
+
+        let on_message_closure = Closure::<dyn FnMut(_)>::new({
+            let new_data_waker = new_data_waker.clone();
+            let read_buffer = read_buffer.clone();
+
+            move |ev: MessageEvent| {
+                let data = js_sys::Uint8Array::new(&ev.data());
+
+                let mut read_buffer = read_buffer.lock().unwrap();
+
+                if read_buffer.len() + data.length() as usize > MAX_MSG_LEN {
+                    // TODO: Should we take as much of the data as we can or drop the entire message?
+
+                    log::warn!(
+                        "Remote is overloading us with messages, dropping {} bytes of data",
+                        data.length()
+                    );
+                    return;
+                }
```

Comment:
This puts us into an inconsistent state, right? We are dropping data but might continue reading on consecutive calls. I imagine this to be hard to debug on the application layer, as it will miss out on a chunk of data. Shouldn't we surface this as an error and fail the whole stream?

Reply:
You mean reset the stream? That is probably a much better idea yeah.

Reply:
> You mean reset the stream?

:+1:

Reply:
I pushed a few commits, let me know if you consider this resolved now. I didn't know how to add tests without not also adding an entire test setup in `wasm-tests`. Let me know what you think.

Reply:
Is it appropriate to send a signal to the other end of the connection when we reset? We do have that `Flag::RESET` available to us, should we be sending this too?

Reply:
We could have the stream [`drop_notifier`](https://github.com/DougAnderson444/rust-libp2p/blob/75e25b58ac2e599122c331283db2ebb71ad9c4f7/misc/webrtc-utils/src/stream.rs#L65) send [`Err(Canceled)`](https://github.com/DougAnderson444/rust-libp2p/blob/75e25b58ac2e599122c331283db2ebb71ad9c4f7/misc/webrtc-utils/src/stream/drop_listener.rs#L81-L83), which will in turn send [`SendingReset`](https://github.com/DougAnderson444/rust-libp2p/blob/75e25b58ac2e599122c331283db2ebb71ad9c4f7/misc/webrtc-utils/src/stream/drop_listener.rs#L94) to the the other end of the channel.

But the stream [`drop_notifier`](https://github.com/DougAnderson444/rust-libp2p/blob/75e25b58ac2e599122c331283db2ebb71ad9c4f7/misc/webrtc-utils/src/stream.rs#L65) is in `stream.rs` and the data channel logic is in `poll_data_channel.rs` so we need to determine the best way to signal this intent from the channel to the stream

Reply:
The error that @thomaseizinger injected bubbles up to the stream [here](https://github.com/DougAnderson444/rust-libp2p/blob/75e25b58ac2e599122c331283db2ebb71ad9c4f7/misc/webrtc-utils/src/stream.rs#L253), where the stream does nothing with it except pass it to the application.

But if we injected a [drop_notifier](https://github.com/DougAnderson444/rust-libp2p/blob/75e25b58ac2e599122c331283db2ebb71ad9c4f7/misc/webrtc-utils/src/stream.rs#L65) handler to send the reset we could let the other end of the connection know:

```rust
Err(e) => {
    log::error!("Error while reading next message: {:?}", e);
    let _ = self
        .drop_notifier
        .take()
        .expect("should be able to take drop_notifier value")
        .send(DropMessage::SendReset);
    return Poll::Ready(Err(e));
}
```

This consumes our `drop_notifier` but the application is going to have to recover from the overloaded pipe anyway

Reply:
See commit https://github.com/libp2p/rust-libp2p/pull/4248/commits/a880251ae03655a29a1462bd5cfa8fc6686bbc9e

Reply:
> The error that @thomaseizinger injected bubbles up to the stream [here](https://github.com/DougAnderson444/rust-libp2p/blob/75e25b58ac2e599122c331283db2ebb71ad9c4f7/misc/webrtc-utils/src/stream.rs#L253), where the stream does nothing with it except pass it to the application.

Well, "does nothing" is correct yes :)
But the consequence of returning the error is that the user has no choice but to drop the stream (they can neither read, write nor close it). Hence, the existing `DropNotifier` would send the reset flag.

Maybe this is too subtle and we should do it more explicitly as you've coded it up? I don't really have a preference :)

Reply:
Let me know what you think @DougAnderson444 !

Reply:
Yes I suppose that's true if the end user understands the `Stream` process, error handles the issue properly, and coded it correctly. I do try to design for those who make mistakes and those who don't know what they are doing yet (perhaps they retry the stream on error, and then the buffer has room again, and they miss packets?) I figure that if the library knows the stream should be reset, the library should reset it and issue that opinion to the user application.

I am still learning how rust libp2p internals work, so even I would have to dig into the code and library to figure out where this error would emerge. Because I don't even see where our application interacts with the `Stream` (honest question, I really don't know). Does the error bubble up to the swarm here?:

```rust
loop {
        match swarm.next().await.unwrap() {
        // ...
}
```

As a newcomer to rust-libp2p, if this is something the application should be handling, we should definitely be including it somewhere in the example. You definitely have me curious how an application would even handle this error now.. @thomaseizinger @mxinden

Reply:
The way I coded it, `overloaded` remains set to `true` and never flicks back to false so the stream is completely unusable :)

Any call to `AsyncRead` or `AsyncWrite` will fail. Most likely, the user uses `?` on their stream operations so whatever future they are in with their stream will fail with `Err`. Thus, the stream goes out of scope and the `DropListener` kicks in and sends the reset.

I fully support your underlying viewpoint, the user should not need to handle something correctly here. The only way of how a user could circumvent the sending of the reset with the above strategy is if they'd leak the stream, i.e. never drop it but keep it alive for the remainder of the program (even though the stream has errored). I am not sure why anybody would do that 😅

Does this make sense? It is definitely a suble interaction of several pieces and I am happy to document it more clearly in the code. At the same time, I think it is quite the clean solution :)

Reply:
The interaction with the stream would be in a `ConnectionHandler` as a result of an inbound or outbound upgrade! :)

Reply:
> Ah, that's why we don't see it in userland -- the stream handled by the protocol author? And most users will be using someone else's protocol (like `Ping`). If I understand what you are saying correctly, a protocol like `Ping` should handle the failed `Stream` and create a new one for us, without the application having to re-dial the peer, yes?

Yes-ish. Whether the stream can just be re-established depends on the protocol. For ping, we will emit an error and create a new one. For gossipsub as well. But after I think 5 failed gossipsub streams, we remove the peer from the mesh.

Re-dialing is something entirely different :)

One connection (i.e. dial) can have many many streams.

FWIW, I'd expect most users to at some point write their own protocol and thus interact with a stream. But that actually shouldn't matter much. The only interface you are given as protocol author is `Async{Read,Write}` and if that one returns an error, there is little you can do. Also not that streams can fail for all kinds of reasons. The remote could send you garbage that you can't deserialise into a your message types for example.

> Your explanation does make sense to me, thank you! You know this system much better than I do, so happy to follow your direction on this. Should we remove the code I injected then?

I'd suggest so yeah, simply because it achieves the same thing with less code. Ideally we'd have a test for it but I imagine this being rather tricky to test.

Reply:
Resolving here in order for the pull request to merge. Thank you for the thorough work @thomaseizinger.